src/azure-cli-core/azure/cli/core/parser.py

```diff
@@ -400,46 +432,50 @@ def _check_value(self, action, value):  # pylint: disable=too-many-statements, t
                 if not error_msg:
                     # parser has no `command_source`, value is part of command itself
-                    error_msg = "{prog}: '{value}' is not in the '{prog}' command group. See '{prog} --help'." \
-                        .format(prog=self.prog, value=value)
-                    if use_dynamic_install.lower() == 'no':
-                        extensions_link = 'https://docs.microsoft.com/en-us/cli/azure/azure-cli-extensions-overview'
-                        error_msg = ("{msg} "
-                                     "If the command is from an extension, "
-                                     "please make sure the corresponding extension is installed. "
-                                     "To learn more about extensions, please visit "
-                                     "{extensions_link}").format(msg=error_msg, extensions_link=extensions_link)
+                    error_msg = "'{value}' is misspelled or not recognized by the system.".format(value=value)
+                az_error = AzCLIError(AzCLIErrorType.CommandNotFoundError, error_msg, command=self.prog)
+
             else:
                 # `command_source` indicates command values have been parsed, value is an argument
                 parameter = action.option_strings[0] if action.option_strings else action.dest
-                error_msg = "{prog}: '{value}' is not a valid value for '{param}'. See '{prog} --help'.".format(
+                error_msg = "{prog}: '{value}' is not a valid value for '{param}'.".format(
                     prog=self.prog, value=value, param=parameter)
                 candidates = difflib.get_close_matches(value, action.choices, cutoff=0.7)
+                az_error = AzCLIError(AzCLIErrorType.ArgumentParseError, error_msg, command=self.prog)
 
-            telemetry.set_user_fault(error_msg)
-            with CommandLoggerContext(logger):
-                logger.error(error_msg)
-            if not caused_by_extension_not_installed:
-                if candidates:
-                    print_args = {
-                        's': 's' if len(candidates) > 1 else '',
-                        'verb': 'are' if len(candidates) > 1 else 'is',
-                        'value': value
-                    }
-                    self._suggestion_msg.append("\nThe most similar choice{s} to '{value}' {verb}:"
-                                                .format(**print_args))
-                    self._suggestion_msg.append('\n'.join(['\t' + candidate for candidate in candidates]))
+            _, params, extension = self._get_failure_recovery_arguments()
+            if candidates:
+                az_error.set_recommendation("Did you mean '{}' ?".format(candidates[0]))
+
+            # recommand a command for user
+            recommender = CommandRecommender(command_name_inferred[3:].strip(), params, extension, error_msg, cli_ctx)
```

Comment:
Aladdin expects the command to be passed unchanged.  Autocorrecting the command requires changing the error type as well.

A few examples to clarify what I mean:
* If `az grup` is corrected to `az group` the new error type should be `MissingRequiredSubcommand` rather than `InvalidSubcommand`
* If `az vim create` is corrected to `az vm create` the new error type should be `MissingRequiredParameters` rather than `InvalidSubcommand`.

Reply:
This will be noted as a work item to improve. Will keep sync with you.

Reply:
> `az vim create` is corrected to `az vm create`

@christopher-o-toole May I ask why the new error type of this case is `MissingRequiredParameters`?

Reply:
@zhoxing-ms 

Running `az vm create` yields the following error:

```
> az vm create
az vm create: error: the following arguments are required: --name/-n, --resource-group/-g
```

Therefore the required parameters `--name` and `--resource-group` are missing in this case.

If we take into account which parameters and arguments the user passed, the logic is more complicated. In terms of getting relevant suggestions however, it suffices to set either `ExpectedArgument` or `MissingRequiredParameters` as the error type assuming that the inferred command is what the user intended.